compiler/memmgt.c

```diff
@@ -34,7 +34,7 @@
 /*
  * Memory is allocated in Blocks of 1024 longs, which form a linked list
  */
-#define BLKLEN ((size_t)1024 * sizeof(long))
+#define BLKLEN ((size_t)768 * sizeof(long))
```

Comment:
Did you find that 4K blocks are too large? Just wondering. In any case, I recommend that you add "// was 1024" so that we know that the original setting was, in case we have to go back to it because of testing.

Reply:
fmemalloc was returning NULL with bigger allocations.

Reply:
Something is definitely fishy here - C86 can't do a 4K (=1024 \* sizeof(long)) far allocation, in all of main memory?! 

As per my other comment, it seems we need to be able to see more of the insides of what allocations are occurring, since I am definitely surprised at not being able to allocate 4K!! All I can think of is that main memory is littered with tons of 65+ byte far allocations from the lowered MAX_NEAR_ALLOC, which you said had to be done.

Overall, it seems managing the toolchain memory might be more complicated than originally thought. I had hoped to be able to keep most small allocations in the near heap, but if we can't since the near heap is super tight, which then causes the far heap to run out of memory when allocating a small 4K, then we might have to move to a more complicated memory management algorithm. In the mean time, I'm glad you've got this working, but it seems we could still easily run out of memory, that things might be fragile, depending on the C source.

What I'm thinking we might have to do would be move to a multi-heap allocation manager. That is, instead of just allocating from the near heap, which sounds like its flat out of space for larger programs (and C86 has lots of strings, so not surprising there), the allocator would allocate larger 8-64k chunks of main memory and use them as "arenas" from which smaller allocations would be allocated from. In this way we can possibly manage tons of small allocations, larger than the near heap in the default data segment, but also not fragment main memory. Lets talk more about all this after I get the malloc visibility enhancements into ELKS.

Thanks for your testing and getting the toolchain back working!

Reply:
It is the first call to fmemalloc(), with 4100 of size, nothing allocated before, it returns NULL, in the very start of x86 execution.

Reply:
> I have a feeling that in the past it used to work (before recent changes in ELKS), but not anymore.

It seemed to me the C86 problem occurred after the fmemalloc was added to it, that it worked fine until you tried to compile chess.c with it and then added the fmemalloc, no?

Reply:
I added fmemalloc in the very beginning of development. The only difference is that in the beginning I was using fmemalloc for all allocations, and now we have the mem.c logic.

Reply:
I see, good to know. I think I'll add a capability in ELKS to be able to monitor malloc and fmemalloc requests and sizes, this should give us a better idea as to what is going on, and also allow for easier tuning. The good news is you have everything working now :)